### DISCUSSION TITLE
Neuromorphic (Dynamic Vision) Sensing: Crazyflie meets Hoverfly!

### DISCUSSION THREAD

**Discussion by simondlevy** [bitcraze/discussions#801] [Show and tell]
As I was discussing in my posts about simulation in UnrealEngine, my current research program involves flight control using biologically-inspired ("neuromorphic") sensors like the iniVation [Dynamic Vision Sensor](https://www.youtube.com/watch?v=ppeCFfMli-E).    I just started my first deck [driver](https://github.com/simondlevy/crazyflie-firmware/blob/master/src/deck/drivers/src/edvs.c) for the eDVS and am looking forward to getting it talking to the Crazyflie. While waiting for my deck [PCB](https://oshpark.com/shared_projects/1nv7LbKh) to arrive, I've been working on a simulation of the DVS.  As shown in this [video](https://youtu.be/4g5wU08A-Sg), you can simulate DVS events by  first-differencing (subtraction) of two successive images captured by the simulated camera.  Thanks to the work of @knmcguire on the Python API for Webots, it was easy to modify the wall-following example to work with the simulated DVS.  I have an undergraduate student starting with me on Monday whose first job will be to see whether the simulated DVS events can be used for further stabilization of the vehicle, as used by [flying insects](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005894) in lieu of an accelerometer / altimeter.  Once we have both working (simulator and deck) I hope to do a blog post on this.

Among many things that excite me about Crazyflie is the way you've incorporated the Arduino-style calls into the Deck API.  I spent several years building my own flight-control [package](https://github.com/simondlevy/Hackflight) around the Arduino API for this kind of research, but switched to Crazyflie recently when I learned about the Bolt 1.1 brushless board (because the sensors I'm working with need more lift than you get with a brushed build).  I figured it was worth giving up the Arduino-style coding in order to get all the features of Crazyflie (esp. hover-in-place and swarms), so I'm thrilled to see that with Crazyflie I can apparently have the best of both worlds! 

Unfortunately, iniVation has discontinued the tiny, five-gram Mini eDVS, but there are lots more opportunities for neuromorphic research with Crazyflie.   For example, the latest generation of [tiny FPGAs](https://tinyfpga.com/) may make it possible to do onboard modeling of Spiking Neural Networks (SNNs), as described [here](https://www.osti.gov/servlets/purl/1649022).

**Comment by krichardsson** [comment]
Hi,
this sounds very exciting! I have been interested in doing something with event cameras on the Crazyflie for years, but I never got around to do it. It ill be interesting to see what you will do!

I'm happy that you find that the Crazyflie is a useful platform, that is what we are hoping and aiming for! 

Keep us posted on your progress, I'd love to hear what is happening in the future!

**Comment by simondlevy** [comment]
Thanks, @krichardsson!  As you probably know, the event cameras (DVS) have a huge bandwdith (~10^5 events per second), requiring a high-speed USB or UART connection.  Thankfully the Teensy line of MCUs supports this, and with Teensy 4.0 I can also use the fast CPU (600MHz) to do essential pre-proccessing on the event stream.  Although the 128x128 retina of the DVS would support all kinds of cool algorithms (e.g., optical flow for tunnel navigation), my current goal is just to have the Teensy compute a simple 2D vector (velocityY, velocityZ) which will get passed along to the Crazyflie's state estimator.  Having found  the Teensy's UART to be more robust than [I^2C](https://forum.pjrc.com/threads/60847-Teensy-4-0-I2C-Slave-issue) for this kind of work, I'm using another of the Teensy UARTs to send the data to the Crazyflie.  (Thanks to your excellent new-deck tutorial, I can even monitor the communications wirelessly in the GUI client!)  Currently I'm just sending a stream of alphabet characters to validate the comms, but obviously I want to send those two velocity values, probably as 16-bit signed ints.  Normally I'd use MSP for this (which I see is already built into the firmware), but looking at the deck drivers I don't see MSP being used for them.  So my current plan is to borrow some ideas from your [GTGPS driver code](https://github.com/simondlevy/crazyflie-firmware/blob/master/src/deck/drivers/src/gtgps.c).  But if you think a different approach might be better, please weigh in!

**Reply by krichardsson** [reply]
I think your approach looks sound, pre-processing on the deck sounds like a good idea :-) Just sending the essential information to the Crazyflie STM reduces the load, processing/filtering the data in the deck (if possible) will be more efficient than transferring it to the Crazyflie and filtering it there. Just getting the data out of the UART will add some load, so minimizing the amount of data to transfer can be beneficial. 

You can also look at [the lighthouse deck for inspiration, we use a UART to transport data from the deck to the Crazyflie](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/lighthouse/lighthouse_core.c#L232-L267). In that case (with 4 base stations) we push 1600 frames (of 12 bytes) per second to the Crazyflie.

I'm not exactly sure, but I assume `MSP` in this context means `CPX`? CPX is a pretty new concept and is nice because it routes data, but it will not be as efficient as a raw UART connection. CPX is used in the [AI-deck driver](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/deck/drivers/src/aideck.c) for instance, but I don't think it is the best chose in your case.

**Comment by simondlevy** [comment]
Cool, thanks @krichardsson!  MSP = [Multwii Serial Protocol.](https://github.com/bitcraze/crazyflie-firmware/blob/e56b7fddf6e6ac9e10d9bacfffb4c4bfcd629219/src/modules/src/msp.c)  The lighthouse deck code looks like what I'm currently doing myself.

**Comment by krichardsson** [comment]
Ah, that explains it! I found some definition on the internet that could have been something like a packet based protocol :-)

**Comment by simondlevy** [comment]
My first Bolt 1.1 [flight](https://www.youtube.com/watch?v=HcN851RfTG8)!

**Reply by knmcguire** [reply]
sorry for the delay! We are in and out of our summer holidays. That is looking good!

**Reply by simondlevy** [reply]
LOL, no apologies necessary -- hope you enjoyed the well-deserved break! 

I'm pretty close to getting DVS-based altitude-hold working in the Webots Crazyflie sim.  As usual, a well-tuned EKF is key.  Here's a comparison between the ground-truth altitude/variometer and the corresponding EKF-filtered values from the simulated DVS:
![Figure_1](https://github.com/bitcraze/discussions/assets/6766537/3143aacc-5b1d-4cd7-89b2-889d8342acf5)

**Comment by simondlevy** [comment]
https://youtu.be/pjU2IdmQWTI

**Reply by krichardsson** [reply]
Nice!

**Comment by simondlevy** [comment]
![20230728_141425](https://github.com/bitcraze/discussions/assets/6766537/cb25086f-3013-480d-8f45-ef688f6cbf63)
![20230728_141430](https://github.com/bitcraze/discussions/assets/6766537/c4dad422-77d4-46fe-a7a5-6913c7ae1fac)
Here is my modification of the Bolt 1.1 3D-printed frame you guys designed, to which I added a little mount for the mini Dynamic Vision   Sensor.

**Comment by simondlevy** [comment]
https://youtu.be/TY3PYL-RsT0

Lateral hold based on simulated DVS works even better than altitude-hold!  When I hit the roll stick pretty hard, it still manages to stabilize.

**Reply by knmcguire** [reply]
that looks very good! soon on the real drone as I can see 😄

**Comment by simondlevy** [comment]
[Sneak Peak!](https://oshpark.com/shared_projects/zoCAFh0l)

**Comment by simondlevy** [comment]
![20230905_181345](https://github.com/bitcraze/discussions/assets/6766537/82df3e17-bd76-44a6-a792-b3897ca69e8d)
Custom DVS deck, top view with Teensy 4.0 and 5V step-up/step-down voltage regulator.

**Comment by simondlevy** [comment]
![20230905_181358](https://github.com/bitcraze/discussions/assets/6766537/a4c89ee7-523a-4184-9cd8-ff6e477982ff)
Deck bottom view, with ESP32 TinyPICO Nano for wifi comms

**Reply by krichardsson** [reply]
Looking good!

**Reply by knmcguire** [reply]
cool!

**Comment by simondlevy** [comment]
LOL, if I had bothered to [RTFM](https://en.wikipedia.org/wiki/RTFM#:~:text=RTFM%20is%20an%20initialism%20and,forum%2C%20software%20documentation%20or%20FAQ.), I would've realized that the Bolt has its own onboard voltage regulator, so if I use a 2s lipo or larger, I don't even need that external regulator on my design!  
Proof of concept:
![20230919_154458](https://github.com/bitcraze/discussions/assets/6766537/81b2f6a3-ba54-40e7-ba51-07d427d792ec)

**Reply by tobbeanton** [reply]
> I would've realized that the Bolt has its own onboard voltage regulator, so if I use a 2s lipo or larger, I don't even need that external regulator on my design!

The DC/DC has a 500mA limit (where ~100mA is used by the Crazyflie) to take into account. Just a heads up.

**Reply by simondlevy** [reply]
Thanks for the heads-up, Tobias.  Based on the prototype in the photo above, it appears that the DC/DC will support enough current draw to power the Teensy, TinyPICO, and DVS.  But I appreciate that this may end up being too much current draw, in which case I'll have to look into a separate powering configuration.

**Comment by simondlevy** [comment]
A [working](https://youtube.com/shorts/mmGofa9p1E0?feature=share) deck prototype!

**Reply by knmcguire** [reply]
That's awesome! Now to see if it flies 😄

**Reply by tobbeanton** [reply]
Well done!

**Reply by matejkarasek** [reply]
Very nice, great work! What si the total mass of the deck and the DVS, including the lense?
We would love to test this out on our Bolt-based [Flapper drone](https://flapper-drones.com/wp/nimbleplus/) one day!

**Comment by simondlevy** [comment]
Thanks, @matejkarasek!  The prototype weighs in at 12.7g, but the new version will probably be a gram or two more (slightly larger, with step-up voltage regulator to power the boards from the battery instead of Bolt VCC.

**Reply by matejkarasek** [reply]
Sweet, too bad the mini eDVS is discontinued...
